File: finder4j-backtrack-examples/src/test/java/com/github/K0zka/finder4j/backtrack/examples/sudoku/SudokuStepFactoryTest.java
package com.github.K0zka.finder4j.backtrack.examples.sudoku;

import java.util.List;

import org.junit.Assert;
import org.junit.Test;

import com.github.K0zka.finder4j.backtrack.examples.sudoku.SudokuState;
import com.github.K0zka.finder4j.backtrack.examples.sudoku.SudokuStep;
import com.github.K0zka.finder4j.backtrack.examples.sudoku.SudokuStepFactory;

public class SudokuStepFactoryTest {
	@Test
	public void produce() {
		Short[][] board = new Short[][] {
				{ 1, 2, 3, 4, 5, 6, 7, 8, 9 },
				{ null, null, null, null, null, null, null, null, null },
				{ null, null, null, null, null, null, null, null, null },
				{ null, null, null, null, null, null, null, null, null },
				{ null, null, null, null, null, null, null, null, null },
				{ null, null, null, null, null, null, null, null, null },
				{ null, null, null, null, null, null, null, null, null },
				{ null, null, null, null, null, null, null, null, null },
				{ null, null, null, null, null, null, null, null, null }
				};
		List<SudokuStep> steps = new SudokuStepFactory().produce(new SudokuState(board));
		Assert.assertFalse(steps.isEmpty());
	}

	@Test
	public void produceColumn() {
		Short[][] board = new Short[][] {
				{ 1, 2, 3, 4, 5, 6, 7, 8, 9 },
				{ null, 9, 8, 7, 5, 6, 4, 3, 2 },
				{ null, null, null, null, null, null, null, null, null },
				{ null, null, null, null, null, null, null, null, null },
				{ null, null, null, null, null, null, null, null, null },
				{ null, null, null, null, null, null, null, null, null },
				{ null, null, null, null, null, null, null, null, null },
				{ null, null, null, null, null, null, null, null, null },
				{ null, null, null, null, null, null, null, null, null }
				};
		List<SudokuStep> steps = new SudokuStepFactory().produce(new SudokuState(board));
		Assert.assertTrue(steps.isEmpty());
	}

	@Test
	public void produceColumn2() {
		Short[][] board = new Short[][] {
				{ 1, 2, 3, 4, 5, 6, 7, 8, 9 },
				{ 6, 9, 8, 7, 1, 3, 4, 5, 2 },
				{ null, null, null, null, null, null, null, null, null },
				{ null, null, null, null, null, null, null, null, null },
				{ null, null, null, null, null, null, null, null, null },
				{ null, null, null, null, null, null, null, null, null },
				{ null, null, null, null, null, null, null, null, null },
				{ null, null, null, null, null, null, null, null, null },
				{ null, null, null, null, null, null, null, null, null }
				};
		List<SudokuStep> steps = new SudokuStepFactory().produce(new SudokuState(board));
		Assert.assertFalse(steps.isEmpty());
	}

}

File: finder4j-backtrack-examples/src/test/kotlin/com/github/k0zka/finder4j/backtrack/examples/sudoku/SudokuStepFactoryTest.kt
package com.github.k0zka.finder4j.backtrack.examples.sudoku

import com.github.k0zka.finder4j.backtrack.examples.sudoku.SudokuState.Companion.notSet
import org.junit.Assert
import org.junit.Test

class SudokuStepFactoryTest {
	
	val n = notSet
	
	@Test
	fun produce() {
		val board = arrayOf(arrayOf(1.toShort(), 2.toShort(), 3.toShort(), 4.toShort(), 5.toShort(), 6.toShort(), 7.toShort(), 8.toShort(), 9.toShort()).toShortArray(),
							arrayOf(n, n, n, n, n, n, n, n, n).toShortArray(),
							arrayOf(n, n, n, n, n, n, n, n, n).toShortArray(),
							arrayOf(n, n, n, n, n, n, n, n, n).toShortArray(),
							arrayOf(n, n, n, n, n, n, n, n, n).toShortArray(),
							arrayOf(n, n, n, n, n, n, n, n, n).toShortArray(),
							arrayOf(n, n, n, n, n, n, n, n, n).toShortArray(),
							arrayOf(n, n, n, n, n, n, n, n, n).toShortArray(),
							arrayOf(n, n, n, n, n, n, n, n, n).toShortArray())
		val steps = SudokuStepFactory().produce(SudokuState(board))
		Assert.assertFalse(steps.isEmpty())
	}

	@Test
	fun produceColumn() {
		val board = arrayOf(arrayOf(1.toShort(), 2.toShort(), 3.toShort(), 4.toShort(), 5.toShort(), 6.toShort(), 7.toShort(), 8.toShort(), 9.toShort()).toShortArray(),
							arrayOf(n, 9.toShort(), 8.toShort(), 7.toShort(), 5.toShort(), 6.toShort(), 4.toShort(), 3.toShort(), 2.toShort()).toShortArray(),
							arrayOf(n, n, n, n, n, n, n, n, n).toShortArray(),
							arrayOf(n, n, n, n, n, n, n, n, n).toShortArray(),
							arrayOf(n, n, n, n, n, n, n, n, n).toShortArray(),
							arrayOf(n, n, n, n, n, n, n, n, n).toShortArray(),
							arrayOf(n, n, n, n, n, n, n, n, n).toShortArray(),
							arrayOf(n, n, n, n, n, n, n, n, n).toShortArray(),
							arrayOf(n, n, n, n, n, n, n, n, n).toShortArray())
		val steps = SudokuStepFactory().produce(SudokuState(board))
		Assert.assertTrue(steps.isEmpty())
	}

	@Test
	fun produceColumn2() {
		val board = arrayOf(arrayOf(1.toShort(), 2.toShort(), 3.toShort(), 4.toShort(), 5.toShort(), 6.toShort(), 7.toShort(), 8.toShort(), 9.toShort()).toShortArray(),
							arrayOf(6.toShort(), 9.toShort(), 8.toShort(), 7.toShort(), 1.toShort(), 3.toShort(), 4.toShort(), 5.toShort(), 2.toShort()).toShortArray(),
							arrayOf(n, n, n, n, n, n, n, n, n).toShortArray(),
							arrayOf(n, n, n, n, n, n, n, n, n).toShortArray(),
							arrayOf(n, n, n, n, n, n, n, n, n).toShortArray(),
							arrayOf(n, n, n, n, n, n, n, n, n).toShortArray(),
							arrayOf(n, n, n, n, n, n, n, n, n).toShortArray(),
							arrayOf(n, n, n, n, n, n, n, n, n).toShortArray(),
							arrayOf(n, n, n, n, n, n, n, n, n).toShortArray())
		val steps = SudokuStepFactory().produce(SudokuState(board))
		Assert.assertFalse(steps.isEmpty())
	}

}
